Strain_Tools/strain/models/strain_geostats.py

```diff
@@ -0,0 +1,374 @@
+        '''
+        try:
+            _model = eval(model_type)
+            self._model = _model()
+        except AttributeError:
+            raise ValueError('Model "{}" has not been implemented'.format(model_type))
+        self._model.setParms(
+            sill = sill,
+            range = rang,
+            use_nugget = [True if nugget is not None else False],
+            nugget = nugget
+        )
+
+    def setPoints(self, xy, data):
+        '''
+        Set the data for kriging.
+
+        Parameters
+        ----------
+        xy: 2D ndarray          XY locations of the input data
+        data: 1 or 2D ndarray   Data to krige. If 2D, will do each dim 
+                                separately
+        '''
+        if data.ndim > 2:
+            raise ValueError('I can only handle 1- or 2-D data inputs')
+        if data.ndim ==2:
+            if data.shape[1] > 2:
+                raise ValueError(
+                        'Input data should be an N x 1 or N x 2 matrix'
+                    )
+        self._xy = xy
+        self._values = np.array(data)
+        self._data_dim = self._values.ndim
+
+    def setGrid(self, XY=None):
+        '''
+        Set the query point grid for kriging
+
+        Parameters
+        ----------
+        XY: 2D ndarray  A 2-D array ordered [x y] of query points. 
+        '''
+        if XY is None:
+            _, _, self._XY = makeGrid(self._grid_inc, self._grid_inc, self._strain_range)
+        else:
+            self._XY = XY
+
+    def krige(self, ktype='ok'):
+        ''' 
+        Interpolate velocities using kriging
+        '''
+        # Create the data covariance matrix
+        SIG = compute_covariance(self._model, self._xy)
+        SIG = SIG + np.sqrt(np.finfo(float).eps)*np.eye(SIG.shape[0])
+
+        # create the data/grid covariance and point-wise terms
+        sig0 = compute_covariance(self._model, self._xy, self._XY); 
+        sig2 = self._model.getSigma00()
+
+        # Do different things, depending on if SK, OK, or UK is desired
+        if ktype == 'sk':
+            Dest, Dsig, lam = simple_kriging(SIG, sig0, self._values, sig2)
+        elif ktype == 'ok':
+            Dest, Dsig, lam, nu = ordinary_kriging(SIG, sig0,self._values, sig2)
+        elif ktype == 'uk':
+            Dest, Dsig, lam = universal_kriging(SIG, sig0,self._values, sig2, self._xy, self._XY)
+        else:
+            raise ValueError('Method "{}" is not implemented'.format(self._flag))
+
+        return Dest, Dsig, lam
+
+    def compute(self, myVelfield):
+        '''Compute the interpolated velocity field'''
+        lon, lat, e, n, se, sn = getVel(myVelfield)
+        xy = np.stack([lon, lat], axis=-1)
+        self.setPoints(xy=xy, data = e)
+        Dest_e, Dsig_e, _ = self.krige()
+        self.setPoints(data = n)
+        Dest_n, Dsig_n, _ = self.krige()
+        
+        # Compute strain rates
+        exx, eyy, exy, rot = strain(self._grid_inc, self._grid_inc, Dest_e, Dest_n)
```

Comment:
we should be precise about units. In gpsgridder, the xinc/yinc to strain() are in km but here they are still in degrees. km is probably better for unitless strain.

Reply:
It's a good point to be careful about units. Strain will be unitless either way, as long as the increment and bounds are defined the same (which they should be anyway).

Reply:
but inside strain_on_regular_grid() we compute gradients in "mm/yr per degree" in this function, which will be different from "mm/yr per km" as used elsewhere.  or am I missing something with the Dest_n variables?  there are no bounds included in the strain_on_regular_grid() function.

Reply:
Ah I see what you're saying... the units on velocity matter. Yes, that's right, which I wasn't thinking about. Probably the best thing to do is to make sure everything is consistent in the calling function, and leave strain_on_regular_grid the same; i.e. have the function assume the units are all consistent.

Reply:
As a follow-up to this point, it would be good to think about what happens over very large regions, I'm thinking western US-size. At this scale distortion from earth curvature is a problem; so what should we do? We could convert velocities to degree/yr and leave all units as degrees, but this does seem a bit odd. Not sure what the right answer is. I guess your spherical version of Delaunay would work fine, but we don't have spherical versions of the other codes. I'm thinking about this because Kaj and I had a USGS proposal to look at all of western US (which got rejected, but still thinking about how it would be done).